Add documentation

beroset opened the pull request:
This includes a previous pull request https://github.com/AsteroidOS/asteroidsyncservice/pull/47 but can be rebased if needed.  It adds extensive developer documentation and updates the README.md file.

beroset committed:
Add SERVICE_NAME to top level CMake file

The SERVICE_NAME will be used to refer to the systemd service file so
that systemd can start the daemon when requested by the plugin.

beroset committed:
Add SERVICE_NAME to config file

The SERVICE_NAME will be used to start the systemd service on behalf of
the plugin.

beroset committed:
Add CMAKE_BINARY_DIR to include locations

We add the CMAKE_BINARY_DIR to the include locations for the
asteroidsyncserviceplugin target to allow it to find the program
configuration file.

beroset committed:
Use SERVICE_NAME from config file

Instead of having the service name hardcoded in the file, we use a macro
defined by the top level CMake file so that we can have different
service files for different targets.

beroset committed:
Fix path in service file

The service file had contained a hardcoded location for the
asteroidsyncserviced daemon, but it should instead use the user-settable
CMAKE_INSTALL_PREFIX to point to the correct location.

beroset committed:
Add a small version of the logo for use in docs

beroset committed:
Add CMake file to build docs

This has a dependency on dot and plantuml.  The dependency on dot is
automatically handled by CMake, but the plantuml url is manually
provided and will result in an error if missing.

beroset committed:
Add the doxygen configuration file

This is the configuration file for Doxygen and includes a few things
that are supplied by CMake.  It also includes a footnote macro that
works for either HTML or PDF output.

beroset committed:
Add introduction text

This is a text description that gives an example of what the software
does and how it interacts with other software.

beroset committed:
Add a page describing the architecture

This is the documentation I wished I had when I started working on this
project.  It describes how this fits in with other components of a
synchronization client and what the interactions are among those
components.

beroset committed:
Add docs to describe how D-Bus is used

This gives a very terse description of D-Bus with links to external
sites and an overview of the various places D-Bus is used within this
system and why.

beroset committed:
Add a glossary of terms

Understanding what the various terms mean for D-Bus and Bluetooth is
helpful in understanding the code.  This provides a brief description
for some important terms.

beroset committed:
Add support for documentation

This creates a new WITH_DOC setting that is off by default (to avoid
dependencies on Doxygen, Dot and plantuml).  If enabled, it supports two
additional targets: make doc and make pdf which generate the HTML and
PDF versions of the documentation, respectively.

beroset committed:
Update README.md for new build system

This updates the README.md to describe how to build it using the new
CMake method and also describes the new doc targets.

MagneFire commented on README.md:
```diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
+
+### Building developer documentation (optional)
+
+To build the documentation for the project in either HTML or PDF format requires an additional option for the "pPrepare the build" step listed above.  The setting is `WITH_DOC` which is set to `OFF` by default, but can be turned on:
```
tTypo here

beroset replied:
tThanks! Fixed.

MagneFire commented on README.md:
```diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
+
+### Building developer documentation (optional)
+
+To build the documentation for the project in either HTML or PDF format requires an additional option for the "pPrepare the build" step listed above.  The setting is `WITH_DOC` which is set to `OFF` by default, but can be turned on:
```
I'm not a native speaker. But this sentence is worded a bit odd (between specifying the formats and the additional option)?

Maybe something like: This project provides developer documentation in the HTML or PDF format. This is disabled by default. One can enable the generation of the documentation by adding an option during the "Prepare the build" steps listed above. To turn on the documentation generation use:

beroset replied:
You're right.  I took your suggestion and trimmed even further.

MagneFire commented on README.md:
````diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
+``` 
 
-Run qmake:
+This builds the software documentation in html format, the software is built in the `build/doc/html` subdirectory.  So to view the documents using Firefox, for example, one can use the command `firefox build/doc/html/index.html` which will load the main page.  Of course any browser can be used, and not just Firefox.
````
Inconsistent use of HTML or html.

beroset replied:
Good catch.  Fixed.

MagneFire commented on README.md:
````diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
+``` 
 
-Run qmake:
+This builds the software documentation in html format, the software is built in the `build/doc/html` subdirectory.  So to view the documents using Firefox, for example, one can use the command `firefox build/doc/html/index.html` which will load the main page.  Of course any browser can be used, and not just Firefox.
````
**So** seems unnecessary here?

beroset replied:
On second thought, the whole paragraph about viewing with Firefox seemed more than needed, so I removed most of the paragraph.

MagneFire commented on README.md:
````diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
+``` 
 
-Run qmake:
+This builds the software documentation in html format, the software is built in the `build/doc/html` subdirectory.  So to view the documents using Firefox, for example, one can use the command `firefox build/doc/html/index.html` which will load the main page.  Of course any browser can be used, and not just Firefox.
````
Do I see a double space before _Of course..._?

beroset replied:
I've reduced the paragraph to a single sentence now.

MagneFire commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+# Introduction {#mainpage} #
+The purpose of a [synchronization client](@ref synchronization_client) for an AsteroidOS watch is to provide a means by which data and commands from the machine, generally, a computer or mobile phone running Linux on which the client runs, to the watch.  Although either a computer or mobile phone are commonly used for this, the rest of this document will refer to "the computer" to mean the comput ...
```
In my opinion, this sounds a bit vague, even though you try to clarify things. The use of _the machine_ makes it more strange?
At first read gives the impression that the _synchronization client_ runs on the watch, which isn't the case.

I would change this to start from an even simpler level.

Something like this maybe?
AsteroidOS synchronizes notifications, weather data, music metadata, etc to the watch using a synchronization client running on a Linux computer or a mobile phone.

beroset replied:
Really good point.  I tried a modification of your suggestion.  Please let me know what you think.

MagneFire commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+# Introduction {#mainpage} #
+The purpose of a [synchronization client](@ref synchronization_client) for an AsteroidOS watch is to provide a means by which data and commands from the machine, generally, a computer or mobile phone running Linux on which the client runs, to the watch.  Although either a computer or mobile phone are commonly used for this, the rest of this document will refer to "the computer" to mean the comput ...
+
+It's important to understand that this current work only provides *half* of the required software, and that a separate synchronization client then uses D-Bus to actually perform the synchronization.
```
This _separate synchronization client_ runs of the watch right?

beroset replied:
No, it runs on the computer.  That is really a terrible couple of paragraphs which I have now rewritten.

MagneFire commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+
+An example may be useful to clarify.  In this example, we trace a single request to turn the volume up via the watch's music app.  From the user's point of view, here's a very simplified sequence diagram.  This assumes that the sofware is already running on the computer, that the watch is [paired](@ref paired), and that the watch is [connected](@ref connected).  The music player shown here isn't  ...
+
+\startuml
```
Very cool! Didn't know this was even possible 🚀

beroset replied:
It is.  I have tested both!

MagneFire commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+\enduml
+
+From the developer's point of view, the story is a little more complex.  Since this is the `androidsyncservice` project, we will only describe the reception of a Bluetooth message and trace as this signal is processed on the computer.  First, the message is received by the computer's Bluetooth hardware.  A low level driver, which depends on the particular hardware used, takes care of this part an ...
+The BlueZ software then translates this incoming message into a [D-Bus](https://www.freedesktop.org/wiki/Software/dbus/) messages.  D-Bus is essentially a message manager that allows Linux-based software packages to talk to other Linux-based software packages.  In this instance, BlueZ handles many of the underlying complexity of Bluetooth for us.  The AsteroidOS uses [Bluetooth Low Energy](@ref b ...
```
First sentence has mixed use of singular and plural construction.
(_a D-Bus messages_)

beroset replied:
Good catch!  Fixed.

MagneFire commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+\enduml
+
+From the developer's point of view, the story is a little more complex.  Since this is the `androidsyncservice` project, we will only describe the reception of a Bluetooth message and trace as this signal is processed on the computer.  First, the message is received by the computer's Bluetooth hardware.  A low level driver, which depends on the particular hardware used, takes care of this part an ...
+The BlueZ software then translates this incoming message into a [D-Bus](https://www.freedesktop.org/wiki/Software/dbus/) messages.  D-Bus is essentially a message manager that allows Linux-based software packages to talk to other Linux-based software packages.  In this instance, BlueZ handles many of the underlying complexity of Bluetooth for us.  The AsteroidOS uses [Bluetooth Low Energy](@ref b ...
```
Use of **The** seems not needed here?
(_The AsteroidOS uses..._)

beroset replied:
Right.

MagneFire commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+\enduml
+
+From the developer's point of view, the story is a little more complex.  Since this is the `androidsyncservice` project, we will only describe the reception of a Bluetooth message and trace as this signal is processed on the computer.  First, the message is received by the computer's Bluetooth hardware.  A low level driver, which depends on the particular hardware used, takes care of this part an ...
+The BlueZ software then translates this incoming message into a [D-Bus](https://www.freedesktop.org/wiki/Software/dbus/) messages.  D-Bus is essentially a message manager that allows Linux-based software packages to talk to other Linux-based software packages.  In this instance, BlueZ handles many of the underlying complexity of Bluetooth for us.  The AsteroidOS uses [Bluetooth Low Energy](@ref b ...
```
_asteroidsyncserviced_ first occurrence of the service with a postfixed with the letter **d**.

beroset replied:
I changed it to *asteroidsyncservice* daemon for clarity.

MagneFire commented on doc/99_glossary.md:
```diff
@@ -0,0 +1,23 @@
+Bluetooth Low Energy is a specialization of the Bluetooth specification that is designed to be used with devices that need to conserve power.  Since AsteroidOS is designed to run on watches that run on batteries, using BLE is a natural fit.  See https://en.wikipedia.org/wiki/Bluetooth_Low_Energy for more details.
+
+### characteristic {#characteristic} ###
+a single data value in BLE that is transferred between devices.  An example is the current volume setting of media player software.
```
This and the next sections start with a lower case. Is this on purpose?

beroset replied:
It was intentional, but I've changed my mind and now all of the definitions are complete sentences.

MagneFire commented on doc/99_glossary.md:
```diff
@@ -0,0 +1,23 @@
+a single data value in BLE that is transferred between devices.  An example is the current volume setting of media player software.
+
+### connected {#connected} ###
+devices which are paired and both ready to send or receive data are connected.  Note that it is possible for devices to be paired but not connected, but they cannot be connected and unpaired.
```
Nit pick. One cannot send and receive at the same time (and/or might be better)?

beroset replied:
I think you're right. I've used your suggested wording.

MagneFire commented on doc/99_glossary.md:
```diff
@@ -0,0 +1,23 @@
+two Bluetooth devices which have already completed pairing are said to be paired.
+
+### pairing {#pairing} ###
+the process by which a computer and a watch are "introduced" to each other so that they can begin to communicate via Bluetooth 
```
Misses a **.**

beroset replied:
I've tried to make all of the entries more consistent.  Now they are all proper sentences.

MagneFire commented on doc/99_glossary.md:
```diff
@@ -0,0 +1,23 @@
+the process by which a computer and a watch are "introduced" to each other so that they can begin to communicate via Bluetooth 
+
+### service {#service} ###
+In the context of BLE, a service can be though of as a collection of related characteristics.
```
_though of_ or _thought of_?

beroset replied:
You're right.  Fixed.

MagneFire reviewed (commented):
Thank you so much for this!
This is a very welcome addition and it helps current and future developers to get started with working on the synchronization client.

Left some comments. I should probably add that some may or may not make sense as I'm not a native speaker. Just ignore my silly mistakes please 😉

beroset committed:
Simplify wording of README.md and add prerequisites

This simplifies and streamlines the wording with many good suggestions
from Magnefire.

beroset committed:
Clarify and reword the introduction

This is somewhat more clear and direct, thanks to many good suggestions
from Magnefire.

beroset committed:
Make entries more consistent and clearer

All entries are not complete sentences and the definition for a
synchronization client is perhaps clearer now.

beroset commented:
Thanks for the thorough reading and review.  I've accepted and (I hope!) improved almost all of your suggestions and I think it's better now as a result.  All of the silly errors were mine, and not yours!

beroset committed:
Add other md files and plantuml configuration

This updates the documentation to add a few missing md files that had
been prepared but not included, and adds the plantuml definition so that
the diagrams are created.

beroset committed:
Add asteroid logo to documentation

FlorentRevest commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+# Introduction {#mainpage} #
+The purpose of a [synchronization client](@ref synchronization_client) is to synchronize data between a device (a computer, phone or tablet) and a watch running AsteroidOS. Although either a computer or mobile phone are commonly used for this, the rest of this document will refer to "the computer" to mean the computer on which the synchronization client runs.
+
+It's important to understand that this current doesn't implement an entire synchronization client but provides relatively portable code on which a sychronization client can be built.  Two such examples are Telescope which runs on Ubuntu Touch, and starship which runs on SailfishOS.
```
s/this current/this project/ ?

s/starship/Starship/

moWerk replied:
Or maybe a `currently` or `this project currently`?

beroset replied:
I think I'll go with "this project."  Thanks!

FlorentRevest commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+
+It's important to understand that this current doesn't implement an entire synchronization client but provides relatively portable code on which a sychronization client can be built.  Two such examples are Telescope which runs on Ubuntu Touch, and starship which runs on SailfishOS.
+
+The kinds of things that are synchronized can include things like the time of day and weather data for the next few days, or commands such as taking a screenshot of the watch face from your phone or turning the volume of music playing on your computer up or down from your watch.
```
s/things like// (avoids the repetition of "things")

beroset replied:
Excellent!  Much better that way.

FlorentRevest commented on doc/01_intro.md:
```diff
@@ -0,0 +1,37 @@
+\enduml
+
+From the developer's point of view, the story is a little more complex.  Since this is the `androidsyncservice` project, we will only describe the reception of a Bluetooth message and trace as this signal is processed on the computer.  First, the message is received by the computer's Bluetooth hardware.  A low level driver, which depends on the particular hardware used, takes care of this part an ...
+The BlueZ software then translates this incoming message into [D-Bus](https://www.freedesktop.org/wiki/Software/dbus/) messages.  D-Bus is essentially a message manager that allows Linux-based software packages to talk to other Linux-based software packages.  In this instance, BlueZ handles many of the underlying complexity of Bluetooth for us.  AsteroidOS uses [Bluetooth Low Energy](@ref ble) or ...
```
s/a message manager that allows Linux-based software packages to talk to other Linux-based software packages/an Inter-Process-Communication system for Linux processes/
(IPC is a fairly standard term)

If you surround code snippets (like MediaService::onCharacteristicChanged) with backticks they should be nicely printed out `in monospace`

The API we use to communicate with media players is actually a freedesktop standard called MPRIS fyi

beroset replied:
I like that edit and will use it.

In Doxygen, code bits such as MediaService::onCharacteristicChanged turn into hyperlinks to that item's documentation.  I had thought that putting them in backticks would disrupt that, but it doesn't, so I'm changing it throughout the documentation. 

I've added a note and a hyperlink to the MPRIS interface specification.

FlorentRevest commented on doc/03_dbus.md:
```diff
@@ -0,0 +1,18 @@
+# Understanding D-Bus {#dbus} #
+This is not a comprehensive tutorial on D-Bus, but an overview of generally what D-Bus is, how it works, and how it is used within `asteroidsyncservice`.  For a more thorough view, see https://www.freedesktop.org/wiki/IntroductionToDBus/
+
+D-Bus is a way for different pieces of software to communicate with each other in a standarized way.  This is how the `asteroidsyncservice` software can, for instance, turn the volume up or down on media player software without needing to know exactly which media player software is being used.\footnote{Although the StarfishOS supports D-Bus, it is not used for this purpose on that platform}
```
Regarding the footnote: actually it is, it's just encapsulated by libwatchfishhttps://github.com/piggz/libwatchfish/blob/master/musiccontroller.cpp#L364

beroset replied:
That's a good clarification.  I will change the footnote to say "Although StarfishOS supports D-Bus, it does not use the MPRIS interface for this purpose on that platform."

FlorentRevest commented on doc/02_architecture.md:
```diff
@@ -0,0 +1,39 @@
+Consult the details of those classes to read about the signals and properties that are available to a QML application.
+
+## Platform support ##
+The QML plugin provides two different ways to automatically start the daemon.  One is via [D-Bus](@ref dbus), which is used by the SailfishOS platform and the Linux desktop platform, and the other is via `upstart` which is used by the Ubuntu Touch platform.
```
Iiuc all these "@ref" are supposed to be accompanied by glossary entries ? Are these entries you forgot to document or do I misunderstand @ref ?

beroset replied:
Yes, each of the `@ref` entries refers to some other anchor in the documentation.  In some cases it refers to a glossary entry and in others, such as `@ref dbus` it refers to a page title.

FlorentRevest commented:
Also I think it would be appropriate to link to https://asteroidos.org/wiki/ble-profiles/ when talking about BLE characteristics etc...

And one big thing I'd expect from this documentation is a clear distinction between "asteroidsyncserviced" the daemon and "asteroidsyncservice" the QML plugin which makes it easy to talk to "asteroidsyncserviced", I didn't find it particularly obvious from a first read.

beroset commented:
> Also I think it would be appropriate to link to https://asteroidos.org/wiki/ble-profiles/ when talking about BLE characteristics etc...

Good point!  Done.
 
> And one big thing I'd expect from this documentation is a clear distinction between "asteroidsyncserviced" the daemon and "asteroidsyncservice" the QML plugin which makes it easy to talk to "asteroidsyncserviced", I didn't find it particularly obvious from a first read.

I changed the last sentence of the first paragraph on the 02_architecture page to read "This project contains two major components: a QML plugin named `asteroidsyncserviceplugin` and a daemon with which it interacts named `asteroidsyncserviced`. " and also made that a little more explicit in the diagram.

beroset committed:
Improve wording on introduction page

This changes the wording for clarity in several places and also adds a
link the AsteroidOS page documenting the BLE services and
characteristics.

beroset committed:
Clarify the two major components, plugin and daemon

This names the two major components and clarifies the structure of the
asteroidsyncservice project.  It also puts all component names in
backtick quotes so they appear consistently in Monospace.

beroset committed:
Put component names in backticks

This puts the component names in backticks so they are formatted
consistently in Monospace and also corrects the footnote about
SailfishOS to remove the incorrect claim that it doesn't use D-Bus for
volume control, and to replace it with the correct statment that it
doesn't use MPRIS for that.

FlorentRevest commented:
Oops I only reviewed the incremental diff this time and it looks like I hit the merge button a bit too fast given that this contained commits about the service file...
